src/content/post/2024-04-05-rdf-shacl-and-kuzu.md

```diff
@@ -0,0 +1,343 @@
+---
+
+The Resource Description Framework (RDF) model, along with property graphs, is one of the most common
+graph data models used in practice. In this post, we will explore how you can work with RDF data in Kùzu
+using [RDFLib](https://rdflib.readthedocs.io/en/stable/) and [pySHACL](https://github.com/RDFLib/pySHACL).
+We will demonstrate how to load this data into Kùzu, write SHACL shape constraints to validate the data,
+and use Kùzu Explorer to visualize the resulting RDF graph.
+
+## Basics of RDF
+
+RDF is a simple and flexible data model[^1], where the data and schema are defined in the form of `<subject, predicate, object>`
+triples. A collection of these triples forms a knowledge graph. Due to the standards around RDF, such as RDFS and OWL, it is also a
+*knowledge representation system* that gives clear meaning to certain vocabulary in triples, especially the predicates/verbs. RDF has its historical roots in
+Semantic Web, which itself has roots in good old-fashioned symbolic AI/knowledge representation and reasoning (KRR).
+
+The main pros of RDF are: (i) modeling complex, irregular domains; (ii) seamless schema and data querying; (iii) data integration
+without transformations; and (iv) automatic logical inference/reasoning[^2]. On the other hand, the main cons of building large RDF databases
+are: (i) performance and scalability; (ii) verbosity; and (iii) inherent challenges of modeling complex domains, such as ensuring
+logical consistency in the database.
+
+Kùzu's [docs](https://docs.kuzudb.com/rdf-graphs/rdf-basics/) and our earlier blog post[^1] on RDF provide a much more detailed explanation on
+the terminology and where RDF graphs are useful, but for the purposes of this post, a brief summary is provided in the table below:
+
+| Term | Description |
+| --- | --- |
+| RDF triple | A statement or fact with three components: `<subject, predicate, object>` |
+| RDF graph | A set of RDF triples |
+| Resource | A thing in the world (e.g., an abstract concept), synonymous with "entity" |
+| IRI | **I**nternationalized **R**esource **I**dentifier: A global identifier for resources, specified as a unicode string |
+| Literal | A value that represents a string, number, or date |
+| RDFS (RDF Schema) | Data-modeling vocabulary for RDF data |
+| OWL (Web Ontology Language) | Computational logic-based language to represent complex knowledge about things, groups of things, and the relations between them |
+| Turtle | A human-readable serialization format for RDF data expressing OWL or RDF schemas |
+
+RDF Schema (RDFS) and OWL are part of larger set of standards to model knowledge, and contain a standardized vocabulary
+to describe the structure of RDF graphs. Some examples are shown below:
+
+- `rdf:type` is used to state that a resource is an instance of a class
+- `rdfs:subClassOf` is used to form class hierarchies
+- `owl:sameAs` is used to identify that two resources are the same
+
+## What is SHACL?
```

Comment:
I think this part should be expanded. It is technically too shallow. As a general writing rule, I think we should always give examples when describing new technical content. So, we should give two examples of SHACL constraints and clarify why these constraints are important. There is only one SHACL constraint in the post and it's very artificial (that the length of a string must be 5). If it's not a lot of work, you can even consider making that a bit more interesting.

Reply:
I think the intent that Paco had was to simply demonstrate how the constraint violations are detected by pySHACL. The example dataset we use is really simple, so one idea to make the constraint a little less artificial could be to check that the age field is an integer when it's in fact incorrectly provided as a float `30.0`. I've update the post to use this, and will create a PR to update the source code with this example as well.

I've also stated that "The example shown above is simple enough for demonstration purposes, but SHACL shapes can be used to define more complex constraints." to indicate that the reader is free to experiment with more complex constraints on their own data (I think we should keep the example as simple as possible to avoid deviating too far from the main point - how to use Kùzu with RDFLib and SHACL).